R/SEIaImIsRD.R

```diff
@@ -0,0 +1,356 @@
+            # names(output) <- c("time", "compartment", "value")
+            object@output <- list("states" = states, "changes" = changes)
+            return(object)
+          })
+ 
+
+#' Calculate the basic reproduction number (\code{R_0}) of the system using the next generation matrix approach.
+#' @seealso \url{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6002118/pdf/main.pdf} mathematical
+#' details of the next generation matrix approach.
+#'
+#' The R0 parameter is given by:
+#' \deqn{R_0 = \rho(FV^{-1})},
+#' where \deqn{F=\frac{\partial F_i{(x_0)}}{\partial x_j}} and \deqn{V=\frac{\partial V_i{(x_0)}}{\partial x_j}}
+#' and \code{rho} represents the spectral radius of a matrix (i.e. the largest absolute of eigenvalue).
+#' The \code{F_i} are the new infections, while the \code{V_i} transfers of infections from one compartment to
+#' another. \code{x_0} is the disease-free equilibrium state.
+#'
+#' @param model A model of an SEIaImIsRD class object with initial_conditions and transmission_parameters set.
+#' @return A SEIaImIsRD class object with R0 value stored
+#' @rdname SEIaImIsRD-class
+#' @aliases R0,SEIaImIsRD-method
+#' @export
+#'
+setMethod("R0", "SEIaImIsRD", function(model) {
+  # get required parameters:
+  S <- model@initial_conditions$S
+  beta <- model@transmission_parameters$beta
+  kappa <- model@transmission_parameters$kappa
+  p_symptom <- model@transmission_parameters$p_symptom
+  gamma <- model@transmission_parameters$gamma
+  mu <- model@transmission_parameters$mu
+  # define matrices F and V:
+  F <- matrix(0, 4, 4)
+  V <- matrix(0, 4, 4)
+  F[1, 2:4] <-  c(S * beta$i_asymptomatic, S * beta$i_mild, S * beta$i_severe)
+  V[1, 1] <- kappa
+  V[2, 1] <- -kappa * (1 - p_symptom$i_mild - p_symptom$i_severe)
+  V[2, 2] <- gamma$i_asymptomatic + mu$i_asymptomatic
+  V[3, 1] <- -kappa * p_symptom$i_mild
+  V[3, 3] <- gamma$i_mild + mu$i_mild
+  V[4, 1] <- -kappa * p_symptom$i_severe
+  V[4, 4] <- gamma$i_severe + mu$i_severe
+  # calculate R0 as the spectral radius for the matrix F x V^(-1):
+  eigVals <- eigen(F %*% (solve(V)))$values
+  model@R0 <- max(abs(eigVals))
+  return(model)
+  })
+
+#' # describeIn SEIaImIsRD Plot the outcome of the ode similuation (for any dataframe)
+#'
+#' @param dataframe A dataframe in long format
+#' @param x The variable in the dataframe to be mapped to the x axis in ggploy aesthetics. Default is set to "time".
+#' @param y The variable in the dataframe to be mapped to the y axis in ggploy aesthetics. Default is set to "value".
+#' @param c The variable in the dataframe to be plotted in different lines with different colours. Default is set to "compartment".
+#'
+#' @return A ggplot
+#' @rdname plot_dataframe
+#' @export
+#' @aliases plot_dataframe,ANY,ANY-method
+#'
+plot_dataframe <- function(dataframe, x = "time", y = "value", c = "compartment") {
+  p <- ggplot(dataframe, aes_string(x = x, y = y)) +
+    geom_line(aes_string(colour = c)) +
+    theme_classic()
+  return(p)
+}
```

Comment:
I'm not sure we need this as it's basically just wrapping a standard ggplot call.

Reply:
Moved to tests and vignettes.